Rescan improvements

Doy-lee opened the pull request:
TLDR:

Before 

```
[+2h37m31.167s] Loaded subsystems in 9451.04s (snl: 7855.79s; ons: 0.00s; 176.0 blks/s; 603.78 kB/s)
```

After

```
[+53m22.656s] Loaded subsystems in 3201.86s (get blks: 20.01s; snl: 3129.03s; ons: 7.54s; 521.7 blks/s; 38.14 MB/s)
```

Rescan stats over here https://pastebin.com/HUdsgJC5 to avoid the 64k PR description limit.

---

- Put block loading into a thread pool, this only really helps for the first half of the chain in the latter halfs we are stuck doing SQL ops which are slow.
- Removes `data_for_serialization` and instead I've added a specialised function to walk the SNL state and directly serialise it to binary. Before we had `snl -> data for serialization -> binary -> lmdb` now we have `snl -> binary -> lmdb`.
- This includes the changes from https://github.com/oxen-io/oxen-core/pull/1786 which cleans up some tables that are no longer used.
- Propagate `skip_verify` into the SQL DB when rescanning to skip validation of the payments to insert into the SQL DB. With similar rational that if the payments are being reconstructed from rescanning, then they were previously validated when they were added to the DB in the first place.
- Defer updating of the `height` in the SQL DB until you hit a height that is an archive height OR you're in the last window of recent blocks which the table will store for small re-orgs. This code is gated by rescanning, so once you have rescanned to the tip, you will go back to committing data as the blocks are generated.
- Fix off-by-one in the blockchain height check on startup which caused the daemon to pop one block off everytime the daemon started
- Fix `sql_db_has` in SNL that checks if the SQL DB has the required historical state to detach their DBs to from using only the recent table. This fixes the issue where you load up the daemon using the new build and it rescans the _earliest_ archival block it has which for me was block 1'000'000.
- In the SNL, profiling reveals that in multiple functions we were needlessly walking over the entire SNL multiple times for different purposes. As we enter the pulse era and later stages, there were more of these list walks, with 1k+ nodes this starts to get expensive. I've trimmed this down to 2 walks. Initially at the start we walk the list and extract pulse information, expired nodes and the block leader in a single loop. Then after processing the block, walk the list again to extract the active and decomm nodes.

We can take this a step further and collect the delta to SNs in the SNL after processing the block and just apply that to the list. I did try this for a bit but it's quite a risky change to introduce as you need to make sure you cover _all_ the cases where the SNL is modified and make sure they get applied to the list afterwards. So I've backed that out and just walk the SNL before and after the block is processed to reduce the risk of bugs here.

- Similarly, forward the metadata extracted from the SNL into the SQL DB, in this case, the list of payable service nodes.
- For the SNL, another hot spot was the calculating of pulse entropy hashes for seeding the pulse quorum. I've remediated this by having a sliding window of blocks cached in the SNL. As blocks are loaded, we seed this sliding window and incrementally feed blocks in as they are processed. This prevents the expensive look-up into LMDB for 12 historical blocks of data for each block.

The worst of all is still the SQL DB. 2 main culprits, constant writing to the SQL DB and walking 2 very large lists. When you payout SN's, you walk the entire payable list to accumulate rewards into `block_payments`. You then walk the entire `block_payments` list to execute SQL queries.

These are both slow and combine together to make it more than 50% of duration of an update step. I have changes that made the SQL DB cache the changes into runtime structures and commit them periodically. Surprisingly, this only increased the rescanning speed by 2x, not what I was suspecting so there is more things to cut when the release schedule is a bit more stable.

They're actually in this commit tree in the history, kept it there and overwritten it with a commit on top with the current dev changes + re-adding the less risky changes on top. Can revisit if needed in the future.

- This also ships a profiler with the build. Instructions in readme, `-D TRACY_ENABLE=ON` to compile with profiling support and then launch the profiler alongside and `oxend` will stream timings to the app.

At some point, probably before release I will have to look at reducing the memory usage of storing the recent state for reorgs. The current stable build has a long-standing bug where the recent states stored for the SNL only serialise the quorums. Any reorgs that the daemon hit would have made it rescan from the last 10k interval on the stable branch. Fixing it now means that we correctly store the last 360 SNL states which seems to hover around ~500MiB of memory and will get worse as the list grows. 

With stable branch and its serialisation bug, the same 360 SNL states occupied around ~128MiB. This is a fairly big jump in usage that operators may not be happy about. We can look at dropping that down as long as it doesn't affect anything. There's a comment saying that it was used for blink (which I suppose will be deprecated) and it's also related to reorg and checkpointing code that I'll need to verify before dropping it down.

In terms of the 10k history checkpoints, the last changes that were committed capped the nodes to storing the last 2 years worth, so storing those SNL states relatively bounded and won't grow ad-infinitum.

One more thing is that now there's a sizeable time increase in startup times to launch of mainnet. That will also have to be addressed. I believe, the time being spent there is deserializing the large number of SNL states from LMDB on startup.

darcys22 committed:
local devnet tweaks

Doy-lee committed:
DRAFT: Tracy integration

Doy-lee committed:
Merge separate SQL DB <feature::ETH_BLS branches into one

Doy-lee committed:
Handle 0 oxen address payout modulus gracefully

Doy-lee committed:
Remove batched_payments_accrued raw and paid

The raw table from my understanding is for pop_blocks. Pop blocks was removed
when we redid the SQL DB to support blockchain detach. This was done to unify
the SNL and SQL DB in that the system has a hard requirement that they
need to pop to the same height and update in lock-step because updating
the SNL has side-effects on the SQL. So if one pops and the other
doesn't then it causes the SNL rewards to go out of sync.

Hence the raw table is actually unused. This was discovered when investigating
the origins of delete block payments in the SQL DB which used to be executed
on pop blocks.

It has an accompanying table, the paid table which was a SQL view of the
accrued rewards that, from my understanding, subtracts the amount each
SN was paid from the accrued table. This had unnecessary levels of indirection,
first a view was created, then a trigger was created to hijack when we insert
into the table such that it subtracted the rewards paid out in the
accrued table.

It's possible to skip this busy-work and layers by just subtracting the amount
directly from the accrued table when the payments are "saved".

Doy-lee committed:
Ensure delayed payments archive/recent tables are dropped on reset

Doy-lee committed:
Ensure deprecated SQL tables are removed

Doy-lee committed:
Defer updating of height until HF19

Doy-lee committed:
Log LMDB size in friendly byte sizes

Doy-lee committed:
Remove unused eth address in batch sn payment

Batch SN payment code path is only use in <=HF20 which is prior to ETH
addresses start getting written to the DB. Hence it only needs to handle Oxen
addresses so we can remove the eth address.

For eth addresses we use block_payments which is a
variant<eth_address,oxen_address>.

Doy-lee committed:
Fix unit tests, batched accrued raw has been removed

Doy-lee committed:
Feed blocks into ONS DB even before HF15

Prior to HF15 the ONS DB still accepts blocks and updates the height and
hash of the last block it read. This ensures that if you terminate the daemon
before HF15 and get into a situation where it needs to rescan the DB it doesn't
trigger the assert that requires all the subsystems have synchronised to
the same height.

Doy-lee committed:
DRAFT: Stage SQL changes in runtime structs and defer commit to DB

Doy-lee committed:
Assign the TX/block hash on load

When we load a block we are given the transaction hash and we also know
the blob size. There's very little reason _not_ to assign these into the
transaction because we are already paying the cost of storing them in the struct
we pay a couple of more instructions to memcpy/assign the size which saves
any downstream code from having to calculate it.

This in the bulk-case of rescanning speeds up block processing by about
2x in the early stages of the chain which is an easy win.

Doy-lee committed:
On rescan, load TXs using a threadpool

Doy-lee committed:
Track the time spent loading blocks

Doy-lee committed:
Remove unused headers and extraneous pulse:: namespacing

Doy-lee committed:
Store the SNL blobs on rescan every 60s

Doy-lee committed:
Remove SNL wasteful intermediate serialisation structs

We had intermediate data structures that we marshalled to `state_serialized` and
`quorums_for_serialization` that had a bunch of code that we marshalled into
before marshalling binary then copied into LMDB.

Instead we can just walk through the state_t object directly and marshall that
into binary which then gets copied into LMDB. This saves a bunch of work that is
unnecessary and means we can get rid of these intermediate data structures.

For backwards compat some of these structures still have to be preserved so
that old nodes can adopt the new binary blobs. After the release for HF20 is
made the following release activating HF21 (the transition to ETH) can delete
all the left-over code.

In terms of performance, this drops storing from 1.3s~ to 800ms. The
bulk of the time is serialising the list of blobs into a final blob which is
also still very wasteful. There are multiple solutions to this which are
documented in the commit as code comments. However, since changing the
periodicity of storing to disk to every 60s on rescan this quickly produces
diminishing gains so we can table that potential future for later.

Doy-lee committed:
Commit the SQL DB to restore old behaviour temporarily whilst ensuring syncing works

Doy-lee committed:
Store SQL data in runtime memory before committing to disk

Doy-lee committed:
Occassionally commit the ONS db even if there are no ONS txs

Doy-lee committed:
Pre-walk the SNL and precompute data to avoid extra SNL list walks

Doy-lee committed:
Pre-calculated the expired nodes to avoid another SNL walk

Doy-lee committed:
Cache the pulse block entropy over time to minimise DB reads

Doy-lee committed:
Revert "Store SQL data in runtime memory before committing to disk"

This reverts commit 65824609b371075d15cf377c6aaaa6885429c66b.

Doy-lee committed:
Revert SQL DB to old implementation, to derisk changes

Doy-lee committed:
Commit SQL height at the end of rescan, speed up list gen

Pulse candidates can be sped up by just storing the pubkey+info pair.
This saves the sort step requiring a lookup into a tree everytime to
sort the object.

Remove the precomputed active snode list, this is only used for
calculating pulse candidates, so instead we push those nodes into the
pulse candidate list.

Finally fix up generate_pulse_quorum2. This is now aptly named
generate_pulse_quorum_with_candidates and generate_pulse_quorum now
thunks into that function. For the SNL it directly calls
generate_pulse_quorum_with_candidates because at the start of the block
it extracts multiple pieces of metadata from the SNL in the singular
loop instead of each function trying to act idempotently and walking the
lists themselves.

For the SQL DB rescanning will skip verification of block payments.
Additionally we only commit the height if we're at an archive height or
in the last window of data that must be kept in the recent table.

Committing a height causes old entries to be pruned from the DB so this
should be safe to defer updating whilst in a rescanning state. In normal
state the DB will continue to commit on each individual block as normal.

Doy-lee committed:
Try load as a new blob first before old

Only on the very first run will miss the branch and need to hit the old
case. This is not particularly a relevant change for release mode
binaries. On debug however, if it's trying to load the old blobs first and it
has migrated, it will fail and throw an exception which can be slow due to it
generating a stack trace.

Doy-lee committed:
Add instructions for Tracy profiler and a cmake command to enable it

Doy-lee committed:
Fix off-by-one in startup detach check, fix sql_db_has not using correct table

Doy-lee committed:
Simplify the blob list, re-add if it shows up in profiler

Doy-lee committed:
Format the code

Doy-lee committed:
Formatting

jagerman commented on CMakeLists.txt:
```diff
@@ -806,6 +806,7 @@ add_library(extra INTERFACE)
 target_link_libraries(extra INTERFACE ${CMAKE_DL_LIBS})
 target_link_libraries(extra INTERFACE Boost::boost)
 target_link_libraries(extra INTERFACE Threads::Threads)
+target_link_libraries(extra INTERFACE Tracy::TracyClient)
```
```suggestion
if(TRACY_ENABLE)
    target_link_libraries(extra INTERFACE Tracy::TracyClient)
endif()
```

Doy-lee replied:
Tracy is meant to compile to nothing if `TRACY_ENABLE` is not set, setting the CMake flag defines the preprocessor definition `TRACY_ENABLE` which causes all the Tracy code in the header and implementation to be compiled out like: https://github.com/wolfpld/tracy/blob/4a5a21cdb08c3b554bf1857b4f2f33d7f7db0207/public/tracy/Tracy.hpp#L19 

We could add a shim `tracy.h` that conditionally includes the real `tracy.h` and or otherwise defines the Tracy macros out ourselves and remove the cost of building the thing from Oxen if you'd like?

jagerman replied:
This was meant to go along with the later comment about avoiding the add_subdirectory, in which case the Tracy::TracyClient target wouldn't be available.

Doy-lee replied:
I've done the change which is what I think you intended. Linking to `TracyClient` is what adds the include paths to allow usage of `<tracy/Tracy.hpp>` in the codebase. So getting rid of it then means I added our own `tracy_shim.h` which handles when `TracyClient` is _not_ linked and `<tracy/Tracy.hpp>` is not available.

I think this is better anyway since the build is already super complicated and we can avoid the meta build-steps incurred from including Tracy since we have a lot of build targets that may or may not work with it out of the box.

jagerman commented on external/CMakeLists.txt:
```diff
@@ -158,3 +157,6 @@ endif()
 add_subdirectory(cpptrace EXCLUDE_FROM_ALL)
+
+option(TRACY_ENABLE "Enable the profiler markers to be captured and displayable with the Tracy Profiler" OFF)
+add_subdirectory(tracy)
```
```suggestion
if(TRACY_ENABLE)
    add_subdirectory(tracy)
endif()
```

Doy-lee replied:
Is done in this commit: https://github.com/oxen-io/oxen-core/pull/1789/commits/360ce9b5d12e76457f1c4c9c8034b675221c4197

Doy-lee committed:
Add tracy_shim.h to avoid compiling Tracy at all

Whether or not TRACY_ENABLE was set, the codebase was going to pull in Tracy
and do some extra steps. This is the intended use case of Tracy and the
way they handled this was by having a TRACY_ENABLE preprocessor define
that compiled out any Tracy code. This however still means that a build
without TRACY_ENABLE is going to process the build steps for Tracy which
_can_ be avoided by providing our own shim.

Normally you link to Tracy which exposes <tracy/Tracy.hpp> and friends
that have macros that you use to mark up the codebase. These macros are
conditionally defined by TRACY_ENABLE. To preserve that functionality in the
case where we don't link to the cmake target, Tracy::TracyClient, I have a stub
header which defines our the Tracy macros we use.

If we compile with Tracy, the shim ends up including the real Tracy headers
which define the proper definition of the Tracy macros.

Doy-lee committed:
Reduce RECENT_KEEP_WINDOW to the abs min required, 65 blocks

Doy-lee committed:
Adhere to formatting linter

Doy-lee committed:
Tweak test to handle new 65 SNL history window

One of the tests chose to generate 60 blocks on the main chain and then
generate 2 checkpoints at a point in history before the 60 blocks were added.
It expects to re-org the main chain with 60 blocks back to the chain with 2
checkpoints.

In order to reorg to the chain with 2 check points, it verifies the chain with a
checkpoint preceeding the the 60 blocks. Since we changed our SNL history window
to store exactly 65 blocks, in combination with the last available checkpoint
interval height, the quorum necessary to verify the checkpoint falls further
than the 65 SNL history window.

This causes the test to fail as that quoum is no longer stored by the SNL. I've
tweaked the to only submit 55 blocks which is a shallow enough depth that the
quorum is still available.

At a high level, this does highlight that the chain cannot may have trouble
supporting re-orgs greater than essentially 60 blocks due to the SNL history
window change. But these 60 blocks would _have_ to be non-checkpointed blocks.
For this to happen the SN network has to be non-compliant for 60 consecutive
blocks.

This can happen if the SN network is down and we fallback to PoW blocks that
require 255 rounds in Pulse and a timeout of 1min, that is (1*255*60) minutes or
255 hours of the network being broken before our SNL history window becomes
insufficient.

Alternatively Pulse with nodes that choose to not generate checkpoints could
reach this situation in 2 hours (assuming 2 mins per block). This may or may not
be an issue.

tewinget committed:
sqlite db transaction during large rescans

Large rescans of the service node list / batch rewards database can take
a long time.  Batching block additions in batches of 100 per database
transaction in the rewards database cuts that time roughly in half.

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -902,7 +941,10 @@ bool BlockchainSQLite::add_delayed_payments(
     log::trace(logcat, "BlockchainSQLite::{} called", __func__);
     try {
-        SQLite::Transaction transaction{db, SQLite::TransactionBehavior::IMMEDIATE};
+        std::unique_ptr<SQLite::Transaction> transaction{nullptr};
```
So this ensures that there's a transaction for adding the delayed payments if there isn't one open at the moment. Can use std::optional instead of unique_ptr?

tewinget replied:
Well I *had* done all `unique_ptr` on the existing transactions.  Just a bit ago I changed them all™ to optional...or so I thought.  Fixing.

Doy-lee reviewed (commented):
Minor suggestion, otherwise LGTM
